### ISSUE TITLE
Emojis render as questionmarks

### ISSUE THREAD

**Issue by boehs** [neovim/neovim#15117]
- `nvim --version`: NVIM v0.6.0-dev+43-g02bf251bb
- Operating system/version: 

Edition	Windows 11 Home
Update	21H2
Installed on	‎7/‎13/‎2021
OS build	22000.65
Experience	Windows Feature Experience Pack 421.17400.45.3

- Terminal name/version: Windows terminal 1.10.1933

### Actual behaviour

![](https://e.reeee.ee/54x0JpE9H.gif)

### Expected behaviour

Emojis render

**Comment by boehs** [comment]
wierd issues with emojis also present in nvim-qt

![](https://e.reeee.ee/54x1vp45q.gif)

**Comment by boehs** [comment]
Issue also observed with `0.5` and using vanilla powershell

**Comment by shadmansaleh** [comment]
I don't think it can be a neovim issue as on rendering isn't done by neovim terminal emulater or gui front-end(nvim-qt) . So it's most likely powershells issue . Try a better terminal emulator if available for windows :D

**Comment by boehs** [comment]
> I don't think it can be a neovim issue as on rendering isn't done by neovim terminal emulater or gui front-end(nvim-qt) . So it's most likely powershells issue . Try a better terminal emulator if available for windows :D

What emulator might you suggest?

**Comment by shadmansaleh** [comment]
> What emulator might you suggest?

I don't use windows . So not sure what's the state of terminals in windows :D . You can search online tp see what's available . Also this might be a font issue too . Maybe the font powershell is using doesn't support emojis . You can check if powershell allows you to change fonts or not .

**Comment by djdv** [comment]
Experiencing this as well in `nvim` and `nvim-qt` for various glyphs, related comment here: https://github.com/neovim/neovim/issues/15020#issuecomment-881533053

**Comment by rktjmp** [comment]
What font are you using? And what fallback fonts?

**Comment by HennyH** [comment]
I can successfully display emojis inside of windows terminal (using Cascadia Code PL as my font https://github.com/microsoft/cascadia-code) when using WSL:

![image](https://user-images.githubusercontent.com/3156577/126103072-740efbc5-88ff-48d0-993f-046ad73b4eaa.png)

However when I use powershell I reproduce the display of emojis as question marks:

![image](https://user-images.githubusercontent.com/3156577/126103166-ec2a2c53-ee12-49af-8239-bb0b835eea19.png)

This suggests to me it isn't so much an issue with Windows Terminal, but with PowerShell?

**Comment by djdv** [comment]
For what it's worth, I use PowerShell on Linux and these render correctly within that on top of WSL.
https://youtu.be/yu7joGOs1Wo

**Comment by boehs** [comment]
> What font are you using? And what fallback fonts?

using `firacode` cc @shadmansaleh & @rktjmp 

I know its not a font issue because emojis render fine outside of nvim

**Comment by rktjmp** [comment]
I wonder if WSL pipes through the (new) ConPTY or something else. Don't have a windows machine so it's hard to poke at, just trying to collect data.

Stuff I would try:

- Does PS need to have some unicode mode enabled? https://akr.am/blog/posts/using-utf-8-in-the-windows-terminal
- Is WT/PS working in UTF8 or something else?
- What is different between how PS and WSL are bootstrapped
- Does SSHing from PS to WSL change anything (I would doubt it).
- Does PS render "normal" unicode stuff OK? Say:
  - u00EB ë
  - u0150 ƛ
- Wonder how it behaves when you type `i<C-v>U1F483<CR>` and `i<C-v>U10000<CR>` 

You can see in the nvim-qt paste that it's getting decoded incorrectly because of the `<200d>` point appears which should actually be attached to the prev/next character (what emoji were you trying to paste?).

Sort of similar behaviour in https://github.com/microsoft/vscode/issues/76842 where characters are both getting mis-rendered or not rendered, related to pty.

Also: https://github.com/vim/vim/issues/2800

> Suppose I try to enter the character U+1F923, rofl. **Windows input is all based around UTF-16 code** units, so what I believe is being sent is 0xD83E 0xDD23; Vim should be decoding this UTF-16 and so obtaining U+1F923, but it seems to be treating the input as Unicode code points instead, U+D83E U+DD23, and so I end up with <d83e><dd23> on screen. (Well, for gVim anyway: on Vim in Command Prompt on codepage 65001—the closest to a Unicode codepage that they have—the first half of the surrogate pair seems to go missing; but I won’t mind too much about that if it’s fixed for gVim.) 

No idea if it is still relevant, that all windows input is UTF16 (just in vim or all windows API?) or what, but might be related.

Not sure if any of that is useful.

**Comment by boehs** [comment]
> * ƛ

Not smart enough to test anything else out but it renders odd Unicode characters like `ƛ` fine

**Comment by jay-patel-cognoai** [comment]
> > I don't think it can be a neovim issue as on rendering isn't done by neovim terminal emulater or gui front-end(nvim-qt) . So it's most likely powershells issue . Try a better terminal emulator if available for windows :D
> 
> What emulator might you suggest?

alacritty ??

**Comment by justinmk** [comment]
> Also: [vim/vim#2800](https://github.com/vim/vim/issues/2800)
> 
> > Suppose I try to enter the character U+1F923, rofl. **Windows input is all based around UTF-16 code** units, so what I believe is being sent is 0xD83E 0xDD23; Vim should be decoding this UTF-16 and so obtaining U+1F923, but it seems to be treating the input as Unicode code points instead, U+D83E U+DD23, and so I end up with  on screen. (Well, for gVim anyway: on Vim in Command Prompt on codepage 65001—the closest to a Unicode codepage that they have—the first half of the surrogate pair seems to go missing; but I won’t mind too much about that if it’s fixed for gVim.)

we skipped vim-patch 8.0.1755, we may need to adapt it.

**Comment by boehs** [comment]
> alacritty

Same issue

**Comment by boehs** [comment]
> For what it's worth, I use PowerShell on Linux and these render correctly within that on top of WSL.
> https://youtu.be/yu7joGOs1Wo

Can confirm no issues in WSL
